### ISSUE TITLE
Motorola 68K "scatter/gather" instructions not handled correctly

### ISSUE THREAD

**Issue by scofield972** [uxmal/reko#367]
**Hi uxmal,**

I have additional issues in mentioned architecture,
This is continue issue from issue number #351.

**issue 1**
dropped me massags:

> Rewriting M68k opcode 'fbnlt' is not supported yet.
> Rewriting M68k opcode 'fbnle' is not supported yet.
> Rewriting M68k opcode 'fneg' is not supported yet.
> Rewriting M68k opcode 'fneg' is not supported yet.

I created example code in the [link ](http://www37.zippyshare.com/v/JH8E1iG4/file.html)

**issue 2**
The following code dropped error as you mentioned in issue #351 :
```
 fmove.d fp3, $-8(a6) 
 move.l $-8(a6),d0
 move.l $-4(a6),d1
 move.l d0, $80400000
 move.l d1, $80400004
```

In general, I think it dropped message like "error: Object reference not set to an instance of an object",
I think this is a problem in "fmove.x" and "fcmp.x" instructions with const operands.

Thanks a lot !

**Comment by uxmal** [comment]
Added the instructions that were mssing. After commit bfdd3195bfe569bf2bcc5baaba1ebdbd2b1becf1, the generated code has improved noticeably. **Issue 2** you mention above will tougher to solve in the `master` branch. The `analysis-branch` will deal with this problem in a better way.

**Comment by scofield972** [comment]
Hi uxmal,

Im sorry my friend,
but, how long time it will be fix ?

I tried to decompile my binary ( as I uploaded it for you in previous message ), and Im got **exceptions** errors before following regular error messages (in "master" and "analysis" branches):

> Rewriting M68k opcode 'fbnlt' is not supported yet.
> Rewriting M68k opcode 'fbnle' is not supported yet.
> Rewriting M68k opcode 'fneg' is not supported yet.
> Rewriting M68k opcode 'fneg' is not supported yet.

**Comment by uxmal** [comment]
Strange: the `master` branch has implemented rewriting for `fbnlt`, `fbnle` and `fneg`. This is how I decompiled your sample:
* Pull the most recent commts from `master`
* Build Reko
* Open the `code.bin` executable using Reko's  `File > Open As..` menu command:
![image](https://cloud.githubusercontent.com/assets/10154465/20456891/fed04f0c-ae7f-11e6-8d1b-46ed75c47927.png)
* In the dialog, specify the values appropriate for the binary fragment:
![image](https://cloud.githubusercontent.com/assets/10154465/20456899/352497fc-ae80-11e6-96a8-37f7d420742a.png)
* When the image is loaded, mark the first byte with `Mark Procedure Entry`:
![image](https://cloud.githubusercontent.com/assets/10154465/20457043/d95ad4ba-ae82-11e6-890a-53c49edfcee2.png)
* Scan, Analyze, and Structure the binary. After these steps, I get:
![image](https://cloud.githubusercontent.com/assets/10154465/20457183/aeaf8212-ae85-11e6-97cb-f97504fdc42d.png)
There are no exceptions or errors, but the C code itself is not correct, because of **issue 2**. 

You can follow progress in the `analysis-development` branch which will resolve **issue 2**. It's hard for me to give a concrete estimate as I work on Reko in my spare time when work allows.

**Comment by scofield972** [comment]
Ok thanks, I will follow issue 2.

**Comment by uxmal** [comment]
I've changed the title to more accurately describe what we've been calling "issue 2".

**Comment by scofield972** [comment]
Ok friend NP,
please update once you will solve it.

**Comment by scofield972** [comment]
Hi friend,
there is progress ?
Thanks !

**Comment by uxmal** [comment]
Let's look at the code sample:
```
fmove.d fp3, $-8(a6) 
move.l $-8(a6),d0
move.l $-4(a6),d1
move.l d0, $80400000
move.l d1, $80400004
```
As I explained earlier, to address this issue requires access to type information, so that when the assignment `fmove.d fp3, $-8(a6)` happens, we know that the frame offset -8 has the type `real64`. This would be translated to:
```
Mem1[fp-8:real64] = fp3
```
If we know that the type of frame offset -8 is `real64`, we can deduce that the subsequent copies `move.l $-8(a6),d0` and `move.l $-4(a6),d1` are slicing a `real64` value:
```
d0_2 = SLICE(Mem1[fp-8:real64], 0, word32)
d1_3 = SLICE(Mem1[fp-8:real64], 4, word32)
```
At this point, d0 and d1 are essentially typeless; all we know is that they are chunks of 32 bits. Together, however, we know that the register pair `d0:d1` is a `real64` because the slices are both addressing the same memory (`Mem0[fp-8:real64]`) and together they "cover" that memory. It's clear we need an analysis/transform that will combine the two assignments above to yield:
```
d0_d1_4 = Mem0[fp-8:real64]
d0_2 = SLICE(d0_d1_4, 0, word32)
d0_3 = SLICE(d0_d1_4, 4, word32)
```
The final two statements are today rewritten to:
```
Mem5[0x80400000:word32] = d0_2
Mem6[0x80400004:word32] = d1_3
```
After the (currently not existing) analysis above, and a propagation step, we would have:
```
Mem5[0x80400000:word32] = SLICE(d0_d1_4, 0, word32)
Mem6[0x80400004:word32] = SLICE(d0_d1_4, 4, word32)
```
Another analysis/transform could look at the two assignments and notice that the original `d0_d1_4` variable is being sliced into two pieces that "cover" it completely, and that the memory locations `0x80400000` and `0x80400004` are adjacent, which makes it likely that they constitute a single value too large for the registers. The end result is:
```
d0_d1_4 = Mem0[fp-8:real64]
d0_2 = SLICE(d0_d1_4, 0, word32)
d0_3 = SLICE(d0_d1_4, 4, word32)
Mem5[0x80400000:word64] = d0_d1_4
```
after removal of dead variables (exists today):
```
d0_d1_4 = Mem0[fp-8:real64]
Mem5[0x80400000:word64] = d0_d1_4
```
after coalescing (exists today):
```
Mem5[0x80400000:word64] = Mem0[fp-8:real64]
```
which is the expected output.

To make all of this possible, the Reko data analysis phase needs to be reorganized so that it happens earlier in the decompilation process. This in turn cannot be done until I've finished the work in the `analysis-development` branch. That branch has two major items left before it can be merged into the `master` branch. 

The first one is dealing with x87 FPU variables properly. This involves making sure that the state of the x87 FPU state is properly propagated between procedures, so that if procedure `proc1` leaves a number on the x87 stack, it is properly registered as returning a floating-point value. This is underway and should hopefully be done by the end of the week (if "real world"  doesn't interfere too much).

The second is handling recursion properly. This hinges on being able to accurately determine, for each procedure involved in recursion, what registers / stack values are modified by it. Doing this turns out to be rather complex. I want to get that done before year's end, but again it's contingent on "real world" issues.

Once those two pieces are in place, the `analysis-development` branch should have the same (or more) capabilities as the `master` branch. I will then merge that branch into `master`. The result will likely change all the `subject`s massively, so I will have to review them to make sure nothing too crazy has happened there. 

Finally, once all the above pieces are in place, I can move the data type analysis earlier in the decompilation process, and add the analysis/transform that discovers `SLICE`s of long values and reconstitutes them into their original assignments.

As you can see there is still some work ahead before I can deliver a fix to this issue. I'm highly motivated to fix it, and I ask you to be patient as I work on this at the pace available given my work and family constraints.

**Comment by uxmal** [comment]
My most recent "push" of commits addresses x87 FPU variables in the `analysis-development` branch. There are 34 unit disabled unit tests remaining, most of which are testing Reko's handling of recursive functions.

**Comment by uxmal** [comment]
The `analysis-development`  branch just received an update that merges separate stack accesses into longer ones, which goes a long way towards addressing this issue.

**Comment by scofield972** [comment]
Hi uxmal,

How are you ?
I found few minutes to do code review about the problem/bug
I saw problem in FPU instruction in translation " $imm(a0)" operand.
I saw that CPU regular instruction translated correctly !
I don't know why in FPU operand translation you choose to return "illegal instruction" (in "get_ea_mode_str" function) instead "IndirectIndexedOperand" such as you did in CPU instruction translation.
This change require to change base and outer variables (maybe 1 more variable) types to "Constant" instead primitive type as you did "uint" :-(
Unfortunately 29 days pass and it did not solve yet.
I think it can be very easy and you can help to do it no much time.

Thanks,

**Comment by uxmal** [comment]
Could you send me the bytes of the FPU instruction that isn't correctly translated? ie hex bytes in big endian order:
00 FF 3a 4E 71 
etc.

**Comment by scofield972** [comment]
Hi uxmal,

Currently I think my binaries (as sent in previous messages) can help you,
You can create unit test with opcodes and debug it,
In addition, I suggest to put focus in "F230 XXXX XXXX XXXX XXXX"(move.x) and "F233 XXXX XXXX XXXX XXXX"(cmp.x) instructions.
Another issue (not critical but require) about implementation real96 bytes (80 bytes used),
I found for you [code](https://github.com/daniel-murray/portsf/blob/master/portsf/ieee80.c) that solve this issue (double to real96 conversion and real96 to double).
Please update,

Thanks,

**Comment by uxmal** [comment]
To speed the process along, please be specific. For instance, name the executable and the address at which the isssues are occurring, or the actual bytes of an instruction, the Reko output, and what you expected to appear.

**Comment by uxmal** [comment]
This has been fixed in the `analysis-development` branch. I hope to release that branch in a week or so.
